### PR TITLE
fix(sync): give the gear junctions a clock so a lost link can come back (#1728)

### LINKED ISSUE #1728: A lost dive_equipment link can never be revived: the junction has no clock
Found while tracing #1720. This is the reason a gear loss is permanent rather
than something the next sync heals, so it turns any single bug that drops a
link into unrecoverable data loss.

`DiveEquipment` has exactly two columns, `dive_id` and `equipment_id`, and a
composite primary key. No `id`, no `created_at`, no `updated_at`, no `hlc`.

Two consequences in the sync merge:

1. In `_applyRemoteDeletions`, `localUpdatedAt` comes from
   `_extractUpdatedAtMillis(local)`, which is null for a `dive_equipment` row.
   Both guards that would normally protect a row, the "edited since last sync"
   check and the "newer than the tombstone" check, are therefore false, so a
   remote tombstone is applied unconditionally. The only remaining defences are
   the pending-entity skip and the same-payload contradiction check.

2. In `_mergeEntity`, when a local tombstone exists and the payload does not
   contradict it, `remoteUpdatedAt == null` takes the `continue`. A gear link
   that has been lost anywhere can therefore never be revived by a live copy
   arriving from a peer.

The code already names the risk. The comment in `_applyRemoteDeletions` notes
that uuid-keyed clockless children regenerate with fresh ids so a stale
tombstone cannot match them, but that "the composite-natural-key junctions
(dive_equipment, equipment_set_items) WOULD match a re-inserted row". The
defence claimed there is the contradiction check, which is same-payload only
and holds for an ordinary save, since `dives.hlc` is stamped before the
tombstones so both clear the same watermark.

It stops holding once a row has actually been lost somewhere.
`_applyRemoteDeletions` re-logs every applied deletion with
`logDeletionIfMissing`, minting a fresh local HLC, so the receiving device
republishes that tombstone in its next changeset. Its data section will not
carry the contradicting live row, because that dive's `dives.hlc` did not
change. From then on the tombstone is uncontradicted on every hop and there is
no `updated_at` for anything to compare it against. Every base import carries
the full deletion log, so a pending-entity skip only defers it a round.

Options, roughly in order of size:

- Give `dive_equipment` (and `equipment_set_items`) an `updated_at`, so the
  existing age guards apply. Schema rung plus a backfill; the junction rides
  the dive today, so the dive's `updated_at` is a defensible backfill value.
- Give the junctions a uuid `id` so they behave like the other clockless
  children a stale tombstone cannot match. Larger change, touches the composite
  key.
- Special-case the junctions in `_applyRemoteDeletions` so a tombstone whose
  parent dive has been touched since is not applied.

Filing rather than fixing because the right answer needs a decision about the
junction's identity, and #1726 removes the local bug that was producing the
losses in the first place.

### PR DESCRIPTION
**Pull request by ericgriffin**
Fixes #1728.

`dive_equipment`, `equipment_set_items` and `dive_plan_equipment` each had two
columns and a composite primary key, and no clock of any kind. That is what
made a single dropped gear link permanent rather than something the next sync
heals, so any one bug that drops a link turned into unrecoverable data loss.

## Both halves of the merge failed on the same missing value

`_applyRemoteDeletions` derives both of its guards from `localUpdatedAt`, and
each is written `localUpdatedAt != null && ...`, so a null collapses both to
`false` in one stroke. A clockless table is not weakly protected against a
stale peer tombstone; it is unprotected.

`_mergeEntity` then computed the remote clock as
`hasUpdatedAt ? _extractUpdatedAtMillis(record) : null` and bailed on null,
which made the revival branch not merely hard to reach but unreachable for
exactly these tables. Deletion was absorbing.

The tombstone also gained strength on every hop. `logDeletionIfMissing`
re-mints it locally with a fresh HLC after applying it, so the receiver
republishes it; the dive's own `hlc` did not change, so the live row is not in
that changeset to contradict it. Each hop stripped the last defence, and every
base import carries the full deletion log, so the pending-entity skip only
deferred it a round.

## The fix

Schema **v207** adds a nullable `updated_at` to all three junctions,
backfilled from the parent each one rides. Nullable so a pre-v207 peer's rows
still parse (`fromJson` casts every non-nullable column, so a NOT NULL column
would throw mid-sync). Declared with a Drift `clientDefault` so every insert
site stamps it without remembering to, and a future call site cannot silently
reintroduce a clockless link. The rung is numbered 207 because 204 through 206
are claimed by branches still open.

The junctions deliberately stay `hasUpdatedAt: false`. That flag picks the
merge **strategy**, not whether a timestamp exists: their payload is the key
itself, so there is nothing to overlay, and routing them through the LWW path
would raise a user-visible conflict card for a gear link. `_mergeEntity` now
reads the remote clock unconditionally; a genuinely clockless entity still
yields null and is unaffected. The flag's doc comment now says so.

## Two more gaps found while wiring it up

**Dive-plan gear never synced at all.** `_recordIdForEntity` had cases for
`diveEquipment` and `equipmentSetItems` but not `divePlanEquipment`, so it fell
through to `record['id']`, which those rows do not have. Every incoming
dive-plan gear row resolved to a null id and was counted as malformed rather
than applied. The serializer had always keyed it as a composite in
`fetchRecord`, `deleteRecord` and `allRecordIds`; only the merge disagreed.

A structural test now drives off the live Drift schema and fails for any synced
table whose primary key `recordIdForEntity` cannot resolve, naming the columns
in the failure message. That is the guard that would have caught this one.

**`equipment_set_items` hand-built its wire map**, in both `_exportEquipmentSetItems`
and `fetchRecord`, from the key pair alone. Left alone, the new column would
never have left the device on one side and never been read on the other.

## Rollout hazard, and what the test found

A peer on an older build sends only the key pair. Drift's `insert` uses
`toColumns(nullToAbsent: true)`, so the absent `updatedAt` falls through to the
`clientDefault` and lands as `now` rather than null. Left alone, that restamps
every gear link as freshly edited on every base import from an old peer, which
flips the new guard the other way: legitimate removals stop applying and pile
up as conflicts for as long as one device stays behind.

The six junction upsert sites (three tables x the single- and batch-record
paths) are now one helper that splits rows on whether the **wire** carried a
clock. Rows that did upsert normally; rows that did not leave an existing row's
`updated_at` untouched, while a genuinely new row still gets the stamp.

## Testing

New `test/core/services/sync/sync_junction_clock_test.dart` and
`test/core/database/migration_v207_junction_clock_test.dart`. Every regression
test was mutation-checked against the pre-fix behaviour: the three that assert
the fix flip red, and the control that asserts a genuine deletion of an older
link still applies stays green, so the guard is discriminating rather than
simply refusing every tombstone.

Two pre-existing tests in `sync_junction_reinsert_test.dart` needed their
fixtures aged. They inserted a membership row with no clock and paired it with
a synthetic `deletedAt: 5000`, which the `clientDefault` now makes newer than
the tombstone. They are stamped before the deletion, which is what they always
meant. `migration_v203`'s `currentSchemaVersion` equality is relaxed to
`greaterThanOrEqualTo` so the newest rung's own test is the only tripwire.

The migration helper is PRAGMA-guarded on the parent's `updated_at` as well as
its own table (an old fixture carries a parent stripped to its id, and
selecting a column that is not there aborts the whole open), and the backfill
runs only for a table whose column the call just added, so the `beforeOpen`
backstop costs three PRAGMAs at steady state rather than three table scans on
every launch.

The compatibility floor is unchanged: a new nullable column is explicitly on
the "do NOT raise" list.

### CODE REVIEW

File: lib/core/services/sync/sync_data_serializer.dart [hunk 1]
```diff
@@ -4577,7 +4556,121 @@ class SyncDataSerializer {
...
+  /// gear link as freshly edited, which flips the new age guard the other way:
+  /// legitimate removals stop applying and pile up as conflicts instead.
+  ///
+  /// So rows are split on whether the WIRE carried a clock. Rows that did
+  /// upsert normally. Rows that did not leave an existing row's `updated_at`
+  /// untouched, while a genuinely new row still gets the `clientDefault`
+  /// stamp on insert.
+  ///
+  /// The provenance pointers are overwritten either way, including with null:
+  /// a peer that deleted the parent clears them on purpose.
```
[the remark this answers is not in the document]
Comment (ericgriffin) [reply lib/core/services/sync/sync_data_serializer.dart]: Agreed, and fixed in 284ccfd. The note now says the provenance pointers (viaEquipmentId, viaSetId) are overwritten only on diveEquipment and divePlanEquipment, and that equipmentSetItems has none: its payload is the key itself, so a clockless row that already exists is ignored. I also added a dive-plan test pinning both halves (the local clock survives an old-peer row, and an explicit null pointer is still applied as a clear).
